compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -187,12 +187,12 @@ pub enum MetadataPosition {
     Last,
 }
 
-// For rlibs we "pack" rustc metadata into a dummy object file. When rustc
-// creates a dylib crate type it will pass `--whole-archive` (or the
-// platform equivalent) to include all object files from an rlib into the
-// final dylib itself. This causes linkers to iterate and try to include all
-// files located in an archive, so if metadata is stored in an archive then
-// it needs to be of a form that the linker will be able to process.
+// For rlibs we "pack" rustc metadata into a dummy object file.
+//
+// Historically it was needed because rustc linked rlibs as whole-archive in some cases.
+// In that case linkers try to include all files located in an archive, so if metadata is stored
+// in an archive then it needs to be of a form that the linker is able to process.
+// Now it's not clear whether metadata still needs to be wrapped into an object file or not.
```

Comment:
Some comments in source code say that you cannot put a random non-object file into the first position in an archive.
However, I wasn't actually able to reproduce this - I tried an archive with a garbage first file on Linux and on windows-gnu targets, with both lld and bfd, and it worked successfully.
@bjorn3 Maybe you know where the relevant code in lld or binutils is? I did some search, but couldn't find it.

I'd like to make a YOLO PR that removes the object file wrapping from metadata, but *still* puts it on the first place in the archive, and run this PR through various targets available on CI.
Maybe in 1-2 weeks.

Reply:
If I recall correctly on macOS non-object files must be at the end. I believe I recall hitting this requirement early on in the development of cg_clif.